---
title: "PombePrimer notes"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Windows and Linux versions:  
Both versions are almost identical, differing only in the primer3 installation and the way primer3 is called in server.R.  

### Installation instructions:  
If you clone the app from the git repository you will get [primer3](https://sourceforge.net/projects/primer3/files/primer3/2.3.7/) with it.  
You will need to install [R](https://cran.r-project.org/) and optionally [RStudio](https://www.rstudio.com/products/rstudio/download/). RStudio is not necessary, but it is well integrated with the Shiny package.  
R packages:  
* Shiny ([CRAN download](https://cran.r-project.org/web/packages/shiny/index.html), [RStudio-Shiny description](https://www.rstudio.com/products/shiny/))  
* Biostrings ([Bioconductor download](https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html))  
Package versions tested with the app:  **Shiny 1.0.5** and **Biostrings 2.44.0** in **R 3.4.3**, calling **primer3 2.3.7**.  


### How to run the app:
You can find general information on how to run a Shiny app in the [RStudio-Shiny website](https://shiny.rstudio.com/articles/running.html).  

To run the app from command line (from Windows or Linux), replace the *~/appdirectory/* below by the app root directory (path where the global.R is stored):  
```
R -e "shiny::runApp(appDir = '~/appdirectory/', launch.browser = TRUE)"
```

To run the app from R, open R, set the working directory to the app root directory, and run:  
```
shiny::runApp()
```

To run the app from RStudio, open any of global.R, ui.R or server.R and press the *Run App* (Run the Shiny application) button in the code editor panel.

### Contents of the repository:  

#### data/ folder:  
* **pombe_genome.RData**: This primer design app uses genomic sequences from 2016, named Schizosaccharomyces_pombe.ASM294v2.30.dna.I.fa.gz (dna.I, dna.II and dna.III for each of the chromosomes), which are stored with CDS coordinates (ch.cds) and gene mapping ID's (namemap) in the data/pombe_genome.RData file.  
* **Rscript_updateVectorSequences.R**: This script is used to update the app with vector data (from vectorsequences.txt), and creates a new vectorData.RData and text to manually update ui.R.  
* **vectorData.RData**: this file contains the R data with the vector sequences, and is loaded from global.R.  
* **vectorsequences.ods, vectorsequences.txt**: these are spreadsheets that contain the vector data to be converted into vectorData.RData with Rscript_updateVectorSequences.R.  

#### ResourcesAndDownloads/ folder:  
It contains the tar.gz and compiled (in the linux version of the app) versions or primer3.2.3.7, which is called in server.R to do the primer design.  

#### www/ folder:  
*www* is a reserved name in Shiny, and contains all the materials that are uploaded from the app (with exception of primer3, which is in ResourcesAndDownloads/).  

* **Cloning_protocol.docx**: a description of the Gibson cloning approach, also displayed in HTML in the app.  
* **pombeprimer_protocol.\***: graphic description of the cloning approach in various formats (.svg, .pdf and .png).  
* **vectorMaps/**: this folder contains all the vector maps available to download from the app, in GenBank (.gb) and GenomeCompiler (.gcproj).  

#### global.R, server.R, ui.R: 
R scripts required by the Shiny package. The can be merged into a single script, but split this way they are more manageable. *Global.R* contains the libraries and data to be loaded when the app launches. *ui.R* codes the user interface, and *server.R* codes all data processing and output generation.  


### Credits:  
This app was designed by [Dr. Ruben Alvarez](https://www.linkedin.com/in/rubenalvarezfernandezcreative/), in the laboratory of [Dr. Juan Mata](http://www.bioc.cam.ac.uk/people/uto/mata), at the Department of Biochemistry in the University of Cambridge (UK).